#496 Feature `JSON.NUMINCRBY`

apoorvyadav1111 opened the pull request:
PR addresses #496 

The changes that are included in the PR.
Added support for JSON.NUMINCRBY. added unit tests and integration tests

All tests passed.

Design Considerations:

Using expr.MustModify() function we should be able to modify the json as per custom requirements. If the changes look good, this PR can be used as a reference for other issues containing logic to update the path such as json.toggle. 

cc:- @JyotinderSingh , @jujiale

apoorvyadav1111 commented:
Hi @JyotinderSingh , @jujiale , PR is ready for review. I have tested extensively for many edge cases. Please review.

JyotinderSingh commented on internal/eval/eval.go:
```diff
@@ -2782,3 +2783,153 @@ func evalSELECT(args []string, store *dstore.Store) []byte {
+				oldVal := jsonData.(int64)
+				jsonData = oldVal + incrInt
+				resultArray = append(resultArray, fmt.Sprintf("%d", jsonData))
+			}
+		default:
+			resultArray = append(resultArray, "null")
+		}
+	} else {
+		// Execute the JSONPath query
+
+		expr.MustModify(jsonData, func(value any) (interface{}, bool) {
+			switch utils.GetJSONFieldType(value) {
+			case utils.NumberType:
+				oldVal := value.(float64)
+				var newVal float64
+				if isIncrFloat {
+					newVal = oldVal + incrFloat
+				} else {
+					newVal = oldVal + float64(incrInt)
+				}
+				resultArray = append(resultArray, formatFloat(newVal, isIncrFloat))
+				return newVal, true
+			case utils.IntegerType:
+				if isIncrFloat {
+					oldVal := value.(float64)
+					newVal := oldVal + incrFloat
+					resultArray = append(resultArray, formatFloat(newVal, isIncrFloat))
+					return newVal, true
+				} else {
+					oldVal := value.(int64)
+					newVal := oldVal + incrInt
+					resultArray = append(resultArray, fmt.Sprintf("%d", newVal))
+					return newVal, true
+				}
+			default:
+				resultArray = append(resultArray, "null")
+				return value, false
+			}
+		})
+	}
```
There is a lot of code duplication here.

You can replace this with:
```go
	if path == defaultRootPath {
		newValue, resultString, modified := incrementValue(jsonData, isIncrFloat, incrFloat, incrInt)
		if modified {
			jsonData = newValue
		}
		resultArray = append(resultArray, resultString)
	} else {
		_, err := expr.Modify(jsonData, func(value any) (interface{}, bool) {
			newValue, resultString, modified := incrementValue(value, isIncrFloat, incrFloat, incrInt)
			resultArray = append(resultArray, resultString)
			return newValue, modified
		})
		if err != nil {
			return diceerrors.NewErrWithMessage("invalid JSONPath")
		}
	}
```

where `incrementValue` is implemented as follows:
```go
func incrementValue(value any, isIncrFloat bool, incrFloat float64, incrInt int64) (interface{}, string, bool) {
	switch utils.GetJSONFieldType(value) {
	case utils.NumberType:
		oldVal := value.(float64)
		var newVal float64
		if isIncrFloat {
			newVal = oldVal + incrFloat
		} else {
			newVal = oldVal + float64(incrInt)
		}
		resultString := formatFloat(newVal, isIncrFloat)
		return newVal, resultString, true
	case utils.IntegerType:
		if isIncrFloat {
			oldVal := float64(value.(int64))
			newVal := oldVal + incrFloat
			resultString := formatFloat(newVal, isIncrFloat)
			return newVal, resultString, true
		} else {
			oldVal := value.(int64)
			newVal := oldVal + incrInt
			resultString := fmt.Sprintf("%d", newVal)
			return newVal, resultString, true
		}
	default:
		return value, "null", false
	}
}
```

apoorvyadav1111 commented on internal/eval/eval.go:
```diff
@@ -2782,3 +2783,153 @@ func evalSELECT(args []string, store *dstore.Store) []byte {
+	} else {
+		// Execute the JSONPath query
+
+		expr.MustModify(jsonData, func(value any) (interface{}, bool) {
```
Thanks for this feedback. Will modify in the next commit for review.

apoorvyadav1111 commented:
Hi @JyotinderSingh , Thanks for the feedback, especially for modularizing the increment logic. I have addressed the comments. Please review and merge if everything looks good.

JyotinderSingh commented:
There's a few merge conflicts. Can you please resolve them?

JyotinderSingh commented:
I have made some changes to the `json_test.go` file to make the testing practices for the JSON tests a little more consistent across the tests. Could you take a look and make the required modifications in your PR? It's just a few cosmetic changes.

apoorvyadav1111 commented:
Hi @JyotinderSingh , sure, I'll update, rebase, and submit PR by tonight.

apoorvyadav1111 commented:
Hi @JyotinderSingh , Please find latest code changes. I have made cosmetic changes to the test cases like you suggested using setup data. Additionally, I did some changes for cleanUp after running a particular test (a small change yet, if we incorporate this in all test cases, we can avoid flakiness created by explicit `Firecommand("del k")` Since you are swamped up, cc'ing other reviewers.
cc: @lucifercr07 @pratikpandey21 
Thanks

JyotinderSingh commented on internal/eval/eval.go:
```diff
@@ -3174,3 +3175,147 @@ func evalSELECT(args []string, store *dstore.Store) []byte {
+			return newVal, resultString, true
+		}
+	default:
+		return value, "null", false
```
Define "null" as a constant

JyotinderSingh commented on internal/eval/eval.go:
```diff
@@ -3174,3 +3175,147 @@ func evalSELECT(args []string, store *dstore.Store) []byte {
+		resultArray = append(resultArray, resultString)
+	} else {
+		// Execute the JSONPath query
+
```
Nit: unnecessary newline

JyotinderSingh commented on integration_tests/commands/json_test.go:
```diff
@@ -841,3 +840,88 @@ func TestJsonARRAPPEND(t *testing.T) {
+	}
+	return elements
+}
+func TestJSONNumIncrBy(t *testing.T) {
```
Can you also add logic to validate the final state of the json values using `JSON.GET`? It will ensure the changes are being applied to the actual values correctly.

JyotinderSingh commented on internal/eval/eval.go:
```diff
@@ -3174,3 +3175,147 @@ func evalSELECT(args []string, store *dstore.Store) []byte {
+
+	resultString := `[` + strings.Join(resultArray, ",") + `]`
+
+	newObj := &object.Obj{
+		Value:        jsonData,
+		TypeEncoding: object.ObjTypeJSON,
+	}
+	exp, ok := dstore.GetExpiry(obj, store)
+
+	var exDurationMs int64 = -1
+	if ok {
+		exDurationMs = int64(exp - uint64(utils.GetCurrentTime().UnixMilli()))
+	}
+	// newObj has bydefault expiry time -1 , we need to set it
+	if exDurationMs > 0 {
+		store.SetExpiry(newObj, exDurationMs)
+	}
+
+	store.Put(key, newObj)
```
I think you may not need this chunk of code.
We can probably just replace this with:

```suggestion
	obj.Value = jsonData
```

JyotinderSingh reviewed (changes requested):
Thanks for addressing the reviews @apoorvyadav1111. The changes mostly look good, left a final set of comments.

apoorvyadav1111 commented:
Hi @JyotinderSingh , Thanks for reviewing PR and helping me to improve the code quality of the JSON.NUMINCRBY command.  I have addressed the review comments, added a test case to check behavior on expire time, added constants for null and JSON.GET check.

JyotinderSingh reviewed (approved):
Thanks for the quick turnaround on the reviews, the changes look great. Approved.
